File: ruby/app/models/simulation.rb
module Simulation
  VELOCITY_ITERATIONS = 8
  POSITION_ITERATIONS = 1

  @@bodies = []

  aabb = B2AABB.new
  aabb.lowerBound.Set 0, 0
  aabb.upperBound.Set WIDTH, HEIGHT

  gravity = B2Vec2.new 0, -10
  do_sleep = true
  @@world = B2World.new aabb, gravity, do_sleep

  def self.add_body(&block)
    @@bodies << Body.new(Body::Configuration.new(&block))
  end

  def self.update
    @@world.Step TIMESTEP, VELOCITY_ITERATIONS, POSITION_ITERATIONS
  end

  def self.bodies
    @@bodies
  end

  def self.world
    @@world
  end

  def self.draw window
    @@bodies.each do |body|
      position = body.position
      x, y = position.x, HEIGHT-position.y
      hw, hh = body.dimensions.x / 2.0, body.dimensions.y / 2.0
      c = body.colour
      window.draw_quad x-hw, y-hh, c, x+hw, y-hh, c, x-hw, y+hh, c, x+hw, y+hh, c
    end
  end
end

File: python/src/simulation.py
from Box2D import *
from body import Body

class Simulation:
    VELOCITY_ITERATIONS = 8
    POSITION_ITERATIONS = 1

    def __init__(self, width, height):
        self.bodies = []
        aabb = b2AABB()
        aabb.lowerBound = (0, 0)
        aabb.upperBound = (width, height)
        gravity = (0, -10)
        do_sleep = True
        self.world = b2World(aabb, gravity, do_sleep)

    def update(self, dt):
        self.world.Step(dt, self.VELOCITY_ITERATIONS,
                        self.POSITION_ITERATIONS)

    def add_body(self, *args, **kw):
        self.bodies.append(Body(self, *args, **kw))
